File: python/pyfory/union.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from __future__ import annotations

from pyfory.resolver import NOT_NULL_VALUE_FLAG, NULL_FLAG
from pyfory.serialization import ENABLE_FORY_CYTHON_SERIALIZATION

if ENABLE_FORY_CYTHON_SERIALIZATION:
    from pyfory.serialization import Serializer
else:
    from pyfory._serializer import Serializer


class Union:
    __slots__ = ("_case_id", "_value")

    def __init__(self, case_id: int, value: object) -> None:
        self._case_id = case_id
        self._value = value

    def case_id(self) -> int:
        return self._case_id

    def value(self) -> object:
        return self._value


class UnionSerializer(Serializer):
    """
    Serializer for generated union classes and typing.Union.

    For generated unions, the payload is:
    | case_id (varuint32) | case_value (Any-style value) |
    """

    __slots__ = (
        "type_resolver",
        "_typing_union",
        "_alternative_types",
        "_alternative_serializers",
        "_case_types",
        "_case_typeinfos",
    )

    def __init__(self, fory, type_, alternative_types):
        super().__init__(fory, type_)
        self.type_resolver = fory.type_resolver
        if isinstance(alternative_types, dict):
            self._typing_union = False
            self._case_types = alternative_types
            self._case_typeinfos = {}
            self._alternative_types = None
            self._alternative_serializers = None
        else:
            self._typing_union = True
            self._alternative_types = alternative_types
            self._case_types = None
            self._case_typeinfos = None
            self._alternative_serializers = []
            for alt_type in alternative_types:
                serializer = fory.type_resolver.get_serializer(alt_type)
                self._alternative_serializers.append((alt_type, serializer))

    def write(self, buffer, value):
        if self._typing_union:
            self._write_typing_union(buffer, value)
            return
        case_id = value.case_id()
        buffer.write_varuint32(case_id)
        typeinfo = self._get_case_typeinfo(case_id)
        self.fory.write_ref(buffer, value._value, typeinfo=typeinfo)

    def read(self, buffer):
        if self._typing_union:
            return self._read_typing_union(buffer)
        case_id = buffer.read_varuint32()
        value = self.fory.read_ref(buffer)
        return self._build_union(case_id, value)

    def xwrite(self, buffer, value):
        if self._typing_union:
            self._xwrite_typing_union(buffer, value)
            return
        case_id = value.case_id()
        buffer.write_varuint32(case_id)
        typeinfo = self._get_case_typeinfo(case_id)
        serializer = typeinfo.serializer
        if serializer.need_to_write_ref:
            if self.fory.ref_resolver.write_ref_or_null(buffer, value._value):
                return
        else:
            if value._value is None:
                buffer.write_int8(NULL_FLAG)
                return
            buffer.write_int8(NOT_NULL_VALUE_FLAG)
        self.type_resolver.write_typeinfo(buffer, typeinfo)
        serializer.xwrite(buffer, value._value)

    def xread(self, buffer):
        if self._typing_union:
            return self._xread_typing_union(buffer)
        case_id = buffer.read_varuint32()
        value = self.fory.xread_ref(buffer)
        return self._build_union(case_id, value)

    def _get_case_typeinfo(self, case_id: int):
        typeinfo = self._case_typeinfos.get(case_id)
        if typeinfo is None:
            case_type = self._case_types.get(case_id)
            if case_type is None:
                raise ValueError(f"unknown union case id: {case_id}")
            typeinfo = self.type_resolver.get_typeinfo(case_type)
            self._case_typeinfos[case_id] = typeinfo
        return typeinfo

    def _build_union(self, case_id: int, value: object):
        if case_id not in self._case_types:
            raise ValueError(f"unknown union case id: {case_id}")
        builder = getattr(self.type_, "_from_case_id", None)
        if builder is None:
            raise TypeError(f"{self.type_} must define _from_case_id for union deserialization")
        return builder(case_id, value)

    def _write_typing_union(self, buffer, value):
        active_index = None
        active_serializer = None

        for i, (alt_type, serializer) in enumerate(self._alternative_serializers):
            if isinstance(value, alt_type):
                active_index = i
                active_serializer = serializer
                break

        if active_index is None:
            raise TypeError(f"Value {value} of type {type(value)} doesn't match any alternative in Union{self._alternative_types}")

        buffer.write_varuint32(active_index)
        active_serializer.write(buffer, value)

    def _read_typing_union(self, buffer):
        stored_index = buffer.read_varuint32()
        if stored_index >= len(self._alternative_serializers):
            raise ValueError(f"Union index out of bounds: {stored_index} (max: {len(self._alternative_serializers) - 1})")
        _, serializer = self._alternative_serializers[stored_index]
        return serializer.read(buffer)

    def _xwrite_typing_union(self, buffer, value):
        active_index = None
        active_serializer = None
        active_type = None

        for i, (alt_type, serializer) in enumerate(self._alternative_serializers):
            if isinstance(value, alt_type):
                active_index = i
                active_serializer = serializer
                active_type = alt_type
                break

        if active_index is None:
            raise TypeError(f"Value {value} of type {type(value)} doesn't match any alternative in Union{self._alternative_types}")

        buffer.write_varuint32(active_index)
        typeinfo = self.type_resolver.get_typeinfo(active_type)
        self.type_resolver.write_typeinfo(buffer, typeinfo)
        active_serializer.xwrite(buffer, value)

    def _xread_typing_union(self, buffer):
        stored_index = buffer.read_varuint32()
        if stored_index >= len(self._alternative_serializers):
            raise ValueError(f"Union index out of bounds: {stored_index} (max: {len(self._alternative_serializers) - 1})")
        typeinfo = self.type_resolver.read_typeinfo(buffer)
        return typeinfo.serializer.xread(buffer)

File: go/fory/union.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"fmt"
	"reflect"
)

// UnionGetter exposes union case metadata for generic serialization.
// Generated union types implement this method.
type UnionGetter interface {
	ForyUnionGet() (uint32, any)
}

// UnionSetter allows setting union case data during deserialization.
// Generated union types implement this method on pointer receivers.
type UnionSetter interface {
	ForyUnionSet(caseId uint32, value any)
}

// UnionCase describes a single union alternative.
type UnionCase struct {
	ID     uint32
	Type   reflect.Type
	TypeID TypeId
}

type unionCaseInfo struct {
	id            uint32
	type_         reflect.Type
	typeID        TypeId
	serializer    Serializer
	needsOverride bool
}

// UnionSerializer is a generic serializer for generated Go unions.
// It relies on union case metadata provided at registration time.
type UnionSerializer struct {
	cases       []unionCaseInfo
	caseByID    map[uint32]*unionCaseInfo
	initialized bool
	initErr     error
}

// NewUnionSerializer creates a generic union serializer for the provided cases.
func NewUnionSerializer(cases ...UnionCase) *UnionSerializer {
	caseInfos := make([]unionCaseInfo, len(cases))
	caseByID := make(map[uint32]*unionCaseInfo, len(cases))
	var initErr error
	for i, c := range cases {
		if c.Type == nil {
			initErr = fmt.Errorf("union case %d has nil type", c.ID)
			continue
		}
		caseInfos[i] = unionCaseInfo{
			id:     c.ID,
			type_:  c.Type,
			typeID: c.TypeID,
		}
		if _, ok := caseByID[c.ID]; ok && initErr == nil {
			initErr = fmt.Errorf("duplicate union case id %d", c.ID)
		}
		caseByID[c.ID] = &caseInfos[i]
	}
	return &UnionSerializer{
		cases:    caseInfos,
		caseByID: caseByID,
		initErr:  initErr,
	}
}

func (s *UnionSerializer) initialize(typeResolver *TypeResolver) error {
	if s.initialized {
		return s.initErr
	}
	if s.initErr != nil {
		return s.initErr
	}
	for i := range s.cases {
		info := &s.cases[i]
		serializer, err := typeResolver.getSerializerByType(info.type_, false)
		if err != nil {
			return err
		}
		info.serializer = serializer
		defaultTypeId := typeResolver.getTypeIdByType(info.type_)
		if defaultTypeId == 0 {
			base := info.type_
			if base.Kind() == reflect.Ptr {
				base = base.Elem()
			}
			defaultTypeId = typeIdFromKind(base)
		}
		info.needsOverride = needsUnionEncodingOverride(info.type_, info.typeID, defaultTypeId)
	}
	s.initialized = true
	return nil
}

// Write is the entry point for union serialization.
func (s *UnionSerializer) Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value) {
	_ = hasGenerics
	switch refMode {
	case RefModeTracking:
		if !value.IsValid() || (value.Kind() == reflect.Ptr && value.IsNil()) {
			ctx.Buffer().WriteInt8(NullFlag)
			return
		}
		refWritten, err := ctx.RefResolver().WriteRefOrNull(ctx.Buffer(), value)
		if err != nil {
			ctx.SetError(FromError(err))
			return
		}
		if refWritten {
			return
		}
	case RefModeNullOnly:
		if !value.IsValid() || (value.Kind() == reflect.Ptr && value.IsNil()) {
			ctx.Buffer().WriteInt8(NullFlag)
			return
		}
		ctx.Buffer().WriteInt8(NotNullValueFlag)
	}
	if writeType {
		typeInfo, err := ctx.TypeResolver().GetTypeInfo(value, true)
		if err != nil {
			ctx.SetError(SerializationErrorf("failed to get type info for union: %v", err))
			return
		}
		ctx.TypeResolver().WriteTypeInfo(ctx.Buffer(), typeInfo, ctx.Err())
	}
	s.WriteData(ctx, value)
}

// WriteData serializes union payload (case_id + case_value).
func (s *UnionSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
	if ctx.HasError() {
		return
	}
	if err := s.initialize(ctx.TypeResolver()); err != nil {
		ctx.SetError(SerializationErrorf("union serializer init failed: %v", err))
		return
	}

	caseID, caseValue, err := getUnionCaseValue(value)
	if err != nil {
		ctx.SetError(SerializationErrorf("union value access failed: %v", err))
		return
	}
	info, ok := s.caseByID[caseID]
	if !ok {
		ctx.SetError(SerializationErrorf("unknown union case id: %d", caseID))
		return
	}

	ctx.Buffer().WriteVaruint32(uint32(caseID))
	if info.needsOverride {
		writeUnionOverrideValue(ctx, info, reflect.ValueOf(caseValue))
	} else {
		ctx.WriteValue(reflect.ValueOf(caseValue), RefModeTracking, true)
	}
}

// Read is the entry point for union deserialization.
func (s *UnionSerializer) Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value) {
	_ = hasGenerics
	err := ctx.Err()
	switch refMode {
	case RefModeTracking:
		refID, refErr := ctx.RefResolver().TryPreserveRefId(ctx.Buffer())
		if refErr != nil {
			ctx.SetError(FromError(refErr))
			return
		}
		if refID < int32(NotNullValueFlag) {
			obj := ctx.RefResolver().GetReadObject(refID)
			if obj.IsValid() {
				value.Set(obj)
			}
			return
		}
	case RefModeNullOnly:
		flag := ctx.Buffer().ReadInt8(err)
		if flag == NullFlag {
			return
		}
	}
	if readType {
		ctx.TypeResolver().ReadTypeInfo(ctx.Buffer(), err)
	}
	s.ReadData(ctx, value)
}

// ReadData deserializes union payload (case_id + case_value).
func (s *UnionSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
	if ctx.HasError() {
		return
	}
	if err := s.initialize(ctx.TypeResolver()); err != nil {
		ctx.SetError(DeserializationErrorf("union serializer init failed: %v", err))
		return
	}
	err := ctx.Err()
	caseID := ctx.Buffer().ReadVaruint32(err)
	if ctx.HasError() {
		return
	}
	info, ok := s.caseByID[caseID]
	if !ok {
		SkipAnyValue(ctx, true)
		if ctx.HasError() {
			return
		}
		ctx.SetError(DeserializationErrorf("unknown union case id: %d", caseID))
		return
	}

	var caseValue any
	if info.needsOverride {
		val, ok := readUnionOverrideValue(ctx, info)
		if !ok {
			return
		}
		caseValue = val
	} else {
		v := reflect.New(info.type_).Elem()
		ctx.ReadValue(v, RefModeTracking, true)
		if ctx.HasError() {
			return
		}
		caseValue = v.Interface()
	}

	setter, setterErr := getUnionSetter(value)
	if setterErr != nil {
		ctx.SetError(DeserializationErrorf("union setter access failed: %v", setterErr))
		return
	}
	setter.ForyUnionSet(caseID, caseValue)
}

// ReadWithTypeInfo deserializes with pre-read type info.
func (s *UnionSerializer) ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value) {
	s.Read(ctx, refMode, false, false, value)
}

func getUnionCaseValue(value reflect.Value) (uint32, any, error) {
	if !value.IsValid() {
		return 0, nil, fmt.Errorf("invalid reflect.Value")
	}
	if value.Kind() == reflect.Interface {
		if value.IsNil() {
			return 0, nil, fmt.Errorf("nil interface value")
		}
		value = value.Elem()
	}
	if value.Kind() == reflect.Ptr {
		if value.IsNil() {
			return 0, nil, fmt.Errorf("nil union pointer")
		}
		if uv, ok := value.Interface().(UnionGetter); ok {
			caseID, caseValue := uv.ForyUnionGet()
			return caseID, caseValue, nil
		}
		value = value.Elem()
	}
	if uv, ok := value.Interface().(UnionGetter); ok {
		caseID, caseValue := uv.ForyUnionGet()
		return caseID, caseValue, nil
	}
	return 0, nil, fmt.Errorf("type %v does not implement fory.UnionGetter", value.Type())
}

func getUnionSetter(value reflect.Value) (UnionSetter, error) {
	if !value.IsValid() {
		return nil, fmt.Errorf("invalid reflect.Value")
	}
	if value.Kind() == reflect.Interface {
		if value.IsNil() {
			return nil, fmt.Errorf("nil interface value")
		}
		value = value.Elem()
	}
	if value.Kind() == reflect.Ptr {
		if value.IsNil() {
			value.Set(reflect.New(value.Type().Elem()))
		}
		if setter, ok := value.Interface().(UnionSetter); ok {
			return setter, nil
		}
	}
	if value.Kind() == reflect.Struct {
		if !value.CanAddr() {
			return nil, fmt.Errorf("union value is not addressable")
		}
		if setter, ok := value.Addr().Interface().(UnionSetter); ok {
			return setter, nil
		}
	}
	return nil, fmt.Errorf("type %v does not implement fory.UnionSetter", value.Type())
}

func needsUnionEncodingOverride(caseType reflect.Type, caseTypeID TypeId, defaultTypeID TypeId) bool {
	if caseTypeID == defaultTypeID {
		return false
	}
	base := caseType
	if base.Kind() == reflect.Ptr {
		base = base.Elem()
	}
	switch base.Kind() {
	case reflect.Int32, reflect.Uint32, reflect.Int64, reflect.Uint64:
		if isNumericEncodingTypeID(caseTypeID) {
			if defaultTypeID == 0 || isNumericEncodingTypeID(defaultTypeID) {
				return true
			}
		}
	}
	return false
}

func isNumericEncodingTypeID(typeID TypeId) bool {
	switch typeID {
	case INT32, VARINT32,
		UINT32, VAR_UINT32,
		INT64, VARINT64, TAGGED_INT64,
		UINT64, VAR_UINT64, TAGGED_UINT64:
		return true
	default:
		return false
	}
}

func writeUnionOverrideValue(ctx *WriteContext, info *unionCaseInfo, value reflect.Value) {
	if value.Kind() == reflect.Interface {
		if !value.IsValid() || value.IsNil() {
			ctx.Buffer().WriteInt8(NullFlag)
			return
		}
		value = value.Elem()
	}
	if !value.IsValid() {
		ctx.Buffer().WriteInt8(NullFlag)
		return
	}
	if value.Kind() == reflect.Ptr {
		if value.IsNil() {
			ctx.Buffer().WriteInt8(NullFlag)
			return
		}
		if isRefKind(value.Elem().Kind()) {
			ctx.SetError(SerializationErrorf("pointer to reference type %s is not allowed", value.Type()))
			return
		}
	}
	refWritten, err := ctx.RefResolver().WriteRefOrNull(ctx.Buffer(), value)
	if err != nil {
		ctx.SetError(FromError(err))
		return
	}
	if refWritten {
		return
	}
	ctx.Buffer().WriteVaruint32Small7(uint32(info.typeID))
	if value.Kind() == reflect.Ptr {
		value = value.Elem()
	}
	writeNumericValueByTypeID(ctx, info.typeID, value)
}

func readUnionOverrideValue(ctx *ReadContext, info *unionCaseInfo) (any, bool) {
	buf := ctx.Buffer()
	refID, refErr := ctx.RefResolver().TryPreserveRefId(buf)
	if refErr != nil {
		ctx.SetError(FromError(refErr))
		return nil, false
	}
	if refID < int32(NotNullValueFlag) {
		obj := ctx.RefResolver().GetReadObject(refID)
		if obj.IsValid() {
			return obj.Interface(), true
		}
		return nil, true
	}

	typeID := TypeId(buf.ReadVaruint32Small7(ctx.Err()))
	if ctx.HasError() {
		return nil, false
	}
	if typeID != info.typeID {
		ctx.SetError(TypeMismatchError(typeID, info.typeID))
		return nil, false
	}

	value, ok := readNumericValueByTypeID(ctx, info.type_, info.typeID)
	if !ok {
		return nil, false
	}
	if value.Kind() == reflect.Ptr {
		ctx.RefResolver().Reference(value)
	}
	return value.Interface(), true
}

func writeNumericValueByTypeID(ctx *WriteContext, typeID TypeId, value reflect.Value) {
	buf := ctx.Buffer()
	switch typeID {
	case INT32:
		buf.WriteInt32(int32(value.Int()))
	case VARINT32:
		buf.WriteVarint32(int32(value.Int()))
	case UINT32:
		buf.WriteUint32(uint32(value.Uint()))
	case VAR_UINT32:
		buf.WriteVaruint32(uint32(value.Uint()))
	case INT64:
		buf.WriteInt64(value.Int())
	case VARINT64:
		buf.WriteVarint64(value.Int())
	case TAGGED_INT64:
		buf.WriteTaggedInt64(value.Int())
	case UINT64:
		buf.WriteUint64(value.Uint())
	case VAR_UINT64:
		buf.WriteVaruint64(value.Uint())
	case TAGGED_UINT64:
		buf.WriteTaggedUint64(value.Uint())
	default:
		ctx.SetError(SerializationErrorf("unsupported union numeric type id: %d", typeID))
	}
}

func readNumericValueByTypeID(ctx *ReadContext, targetType reflect.Type, typeID TypeId) (reflect.Value, bool) {
	buf := ctx.Buffer()
	base := targetType
	isPtr := false
	if base.Kind() == reflect.Ptr {
		isPtr = true
		base = base.Elem()
	}
	var v reflect.Value
	switch base.Kind() {
	case reflect.Int32:
		var value int32
		switch typeID {
		case INT32:
			value = buf.ReadInt32(ctx.Err())
		case VARINT32:
			value = buf.ReadVarint32(ctx.Err())
		default:
			ctx.SetError(DeserializationErrorf("unsupported union int32 type id: %d", typeID))
			return reflect.Value{}, false
		}
		if ctx.HasError() {
			return reflect.Value{}, false
		}
		v = reflect.New(base).Elem()
		v.SetInt(int64(value))
	case reflect.Uint32:
		var value uint32
		switch typeID {
		case UINT32:
			value = buf.ReadUint32(ctx.Err())
		case VAR_UINT32:
			value = buf.ReadVaruint32(ctx.Err())
		default:
			ctx.SetError(DeserializationErrorf("unsupported union uint32 type id: %d", typeID))
			return reflect.Value{}, false
		}
		if ctx.HasError() {
			return reflect.Value{}, false
		}
		v = reflect.New(base).Elem()
		v.SetUint(uint64(value))
	case reflect.Int64:
		var value int64
		switch typeID {
		case INT64:
			value = buf.ReadInt64(ctx.Err())
		case VARINT64:
			value = buf.ReadVarint64(ctx.Err())
		case TAGGED_INT64:
			value = buf.ReadTaggedInt64(ctx.Err())
		default:
			ctx.SetError(DeserializationErrorf("unsupported union int64 type id: %d", typeID))
			return reflect.Value{}, false
		}
		if ctx.HasError() {
			return reflect.Value{}, false
		}
		v = reflect.New(base).Elem()
		v.SetInt(value)
	case reflect.Uint64:
		var value uint64
		switch typeID {
		case UINT64:
			value = buf.ReadUint64(ctx.Err())
		case VAR_UINT64:
			value = buf.ReadVaruint64(ctx.Err())
		case TAGGED_UINT64:
			value = buf.ReadTaggedUint64(ctx.Err())
		default:
			ctx.SetError(DeserializationErrorf("unsupported union uint64 type id: %d", typeID))
			return reflect.Value{}, false
		}
		if ctx.HasError() {
			return reflect.Value{}, false
		}
		v = reflect.New(base).Elem()
		v.SetUint(value)
	default:
		ctx.SetError(DeserializationErrorf("unsupported union numeric kind: %v", base.Kind()))
		return reflect.Value{}, false
	}

	if !isPtr {
		return v, true
	}
	ptr := reflect.New(base)
	ptr.Elem().Set(v)
	return ptr, true
}

func isRefKind(kind reflect.Kind) bool {
	switch kind {
	case reflect.Ptr, reflect.Map, reflect.Slice, reflect.Interface:
		return true
	default:
		return false
	}
}
